<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Usage &mdash; scGO 1.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=f2a433a1"></script>
        <script src="_static/doctools.js?v=888ff710"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Run examples" href="run_examples.html" />
    <link rel="prev" title="Data preprocessing" href="data_preprocessing.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            scGO
          </a>
              <div class="version">
                1.0
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="data_preprocessing.html">Data preprocessing</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#source-code">Source code</a></li>
<li class="toctree-l2"><a class="reference internal" href="#training-cell-type-annotation-model">Training cell type annotation model</a></li>
<li class="toctree-l2"><a class="reference internal" href="#train-scgo-model-using-baron-dataset">Train scGO model using Baron dataset.</a></li>
<li class="toctree-l2"><a class="reference internal" href="#training-regression-model-to-predict-a-continuous-cell-status">Training regression model to predict a continuous cell status</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="run_examples.html">Run examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="help.html">Getting Help</a></li>
<li class="toctree-l1"><a class="reference internal" href="cite.html">Citing scGO</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">scGO</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Usage</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/usage.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="usage">
<h1>Usage<a class="headerlink" href="#usage" title="Link to this heading"></a></h1>
<section id="source-code">
<h2>Source code<a class="headerlink" href="#source-code" title="Link to this heading"></a></h2>
<p>The source code and data processing scripts are available on <a class="reference external" href="https://github.com/yulab2021/scGO">GitHub</a>. You can download them by using the git clone command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">git</span> <span class="n">clone</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">yulab2021</span><span class="o">/</span><span class="n">scGO</span><span class="o">.</span><span class="n">git</span>
</pre></div>
</div>
<p>In the provided repository, the pretrained models are located under the <code class="docutils literal notranslate"><span class="pre">./models</span></code> directory, and the data processing scripts and the main script are located under the <code class="docutils literal notranslate"><span class="pre">./scripts</span></code> directory. The demo data is located under the <code class="docutils literal notranslate"><span class="pre">./demo</span></code> directory.:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>scGO
├── data
│   ├── goa_human.gaf.zip
│   └── TF_annotation_hg38.demo.tsv
├── demo
│   ├── baron_data.csv
│   ├── baron_data_filtered.csv
│   ├── baron_data_filtered.predicted.csv
│   ├── baron_meta_data.csv
│   ├── baron_meta_data_senescence_score.csv
│   ├── baron_meta_data_senescence_score.predicted.csv
│   ├── feature
│   ├── gene_TF_dict
│   ├── gene_to_TF_transform_matrix
│   ├── goa_human.gaf.zip
│   ├── GO_mask
│   ├── GO_TF_mask
│   ├── test_data.csv
│   ├── TF_annotation_hg38.demo.tsv
│   ├── TF_gene_dict
│   └── TF_mask
├── models
│   ├── scGO.demo.pkl
│   └── scGO.senescence_score.demo.pkl
├── README.md
├── results_reproduce
├── scGO.yml
└── scripts
    ├── data_processing.py
    ├── __init__.py
    ├── models.py
    ├── __pycache__
    │   ├── models.cpython-38.pyc
    │   └── utils.cpython-38.pyc
    ├── rds_to_csv.r
    ├── scGO.py
    ├── test.ipynb
    └── utils.py
</pre></div>
</div>
</section>
<section id="training-cell-type-annotation-model">
<h2>Training cell type annotation model<a class="headerlink" href="#training-cell-type-annotation-model" title="Link to this heading"></a></h2>
</section>
<section id="train-scgo-model-using-baron-dataset">
<h2>Train scGO model using <a class="reference external" href="https://www.cell.com/cell-systems/fulltext/S2405-4712(16)30266-6?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS2405471216302666%3Fshowall%3Dtrue">Baron dataset</a>.<a class="headerlink" href="#train-scgo-model-using-baron-dataset" title="Link to this heading"></a></h2>
<p>The Baron dataset covers a range of cell types found in the pancreas, including acinar cells, activated stellate cells, alpha cells, beta cells, delta cells, ductal cells, endothelial cells, epsilon cells, gamma cells, macrophages, mast cells, quiescent stellate cells, and schwann cells. Baron dataset is available at <a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE84133">GSE84133</a>. In this demo, a subsets of the the baron dataset was taken for demonstration purposes due to the large size of the original datasets. The demo dataset was located under <code class="docutils literal notranslate"><span class="pre">./demo/</span></code> directory.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>demo
├── baron_data.csv
├── baron_meta_data.csv
├── goa_human.gaf
└── TF_annotation_hg38.demo.tsv
</pre></div>
</div>
<p><strong>1. Data normalization and filtering</strong></p>
<p>Normalize data and filter out genes that expressed in fewer cells using script <code class="docutils literal notranslate"><span class="pre">data_processing.py</span></code> with command <code class="docutils literal notranslate"><span class="pre">norm_and_filter</span></code>. The number of genes to be retained can be customized by using the argument <code class="docutils literal notranslate"><span class="pre">--num_genes</span></code>. The default value is 2000. The normalized and filtered data will be saved in the <code class="docutils literal notranslate"><span class="pre">./demo/</span></code> folder.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">data_processing</span><span class="o">.</span><span class="n">py</span> <span class="n">norm_and_filter</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">gene_expression_matrix</span> <span class="n">GENE_EXPRESSION_MATRIX</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">num_genes</span> <span class="n">NUM_GENES</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">output</span> <span class="n">OUTPUT</span><span class="p">]</span>
<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
    <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
    <span class="o">--</span><span class="n">gene_expression_matrix</span> <span class="n">GENE_EXPRESSION_MATRIX</span>       <span class="n">Path</span> <span class="n">to</span> <span class="n">the</span> <span class="nb">input</span> <span class="n">data</span>
    <span class="o">--</span><span class="n">num_genes</span> <span class="n">NUM_GENES</span>      <span class="n">Number</span> <span class="n">of</span> <span class="n">top</span> <span class="n">genes</span> <span class="n">to</span> <span class="n">keep</span>
    <span class="o">--</span><span class="n">output</span> <span class="n">OUTPUT</span>        <span class="n">Path</span> <span class="n">to</span> <span class="n">the</span> <span class="n">output</span> <span class="n">data</span>

<span class="n">expample</span><span class="p">:</span>
<span class="n">python</span> <span class="n">scripts</span><span class="o">/</span><span class="n">data_processing</span><span class="o">.</span><span class="n">py</span> <span class="n">norm_and_filter</span> <span class="o">--</span><span class="n">gene_expression_matrix</span> <span class="n">demo</span><span class="o">/</span><span class="n">baron_data</span><span class="o">.</span><span class="n">csv</span> <span class="o">--</span><span class="n">num_genes</span> <span class="mi">2000</span> <span class="o">--</span><span class="n">output</span> <span class="n">demo</span><span class="o">/</span><span class="n">baron_data_filtered</span><span class="o">.</span><span class="n">csv</span>
</pre></div>
</div>
<p><strong>2. Building network connections</strong></p>
<p>Build network connections between gene layer, TF layer and GO layer according to GO annotations and TF annotations using script <code class="docutils literal notranslate"><span class="pre">data_processing.py</span></code> with command <code class="docutils literal notranslate"><span class="pre">build_network</span></code>. The human GO annotation used in this study is downloaded from the <a class="reference external" href="https://doi.org/10.5281/zenodo.7504797">Gene Ontology knowledgebase</a>. The connections between genes and TFs are builded according to the DAP-seq TF annotation data. The DAP-seq data is downloaded from the <a class="reference external" href="https://remap2022.univ-amu.fr/">Remap database</a>. The demo data offers a subset of the processed DAP-seq file for illustrative purposes.  The full processed DAP-seq file has been uploaded to <a class="reference external" href="https://drive.google.com/file/d/1VPSDyNbs4lBITm2VoPD2eJ3BZGcdkrdC/view?usp=drive_link">google drive</a>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">data_processing</span><span class="o">.</span><span class="n">py</span> <span class="n">build_network</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">gene_expression_matrix</span> <span class="n">GENE_EXPRESSION_MATRIX</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">GO_annotation</span> <span class="n">GO_ANNOTATION</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">TF_annotation</span> <span class="n">TF_ANNOTATION</span><span class="p">]</span>
<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
    <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
    <span class="o">--</span><span class="n">gene_expression_matrix</span> <span class="n">GENE_EXPRESSION_MATRIX</span>     <span class="n">Path</span> <span class="n">to</span> <span class="n">the</span> <span class="nb">input</span> <span class="n">data</span>
    <span class="o">--</span><span class="n">GO_annotation</span> <span class="n">GO_ANNOTATION</span>         <span class="n">Path</span> <span class="n">to</span> <span class="n">the</span> <span class="n">GO</span> <span class="n">annotation</span> <span class="n">file</span>
    <span class="o">--</span><span class="n">TF_annotation</span> <span class="n">TF_ANNOTATION</span>         <span class="n">Path</span> <span class="n">to</span> <span class="n">the</span> <span class="n">TF</span> <span class="n">annotation</span> <span class="n">file</span>

<span class="n">example</span><span class="p">:</span>
<span class="n">python</span> <span class="n">scripts</span><span class="o">/</span><span class="n">data_processing</span><span class="o">.</span><span class="n">py</span> <span class="n">build_network</span> <span class="o">--</span><span class="n">gene_expression_matrix</span> <span class="n">demo</span><span class="o">/</span><span class="n">baron_data_filtered</span><span class="o">.</span><span class="n">csv</span> <span class="o">--</span><span class="n">GO_annotation</span> <span class="n">demo</span><span class="o">/</span><span class="n">goa_human</span><span class="o">.</span><span class="n">gaf</span>  <span class="o">--</span><span class="n">TF_annotation</span> <span class="n">demo</span><span class="o">/</span><span class="n">TF_annotation_hg38</span><span class="o">.</span><span class="n">demo</span><span class="o">.</span><span class="n">tsv</span>
</pre></div>
</div>
<p><strong>3. Model training</strong></p>
<p>To train the scGO model using your own dataset from scratch, you can run the <code class="docutils literal notranslate"><span class="pre">scGO.py</span></code> script with the command <code class="docutils literal notranslate"><span class="pre">train</span></code>. scGO accepts both single-cell gene expression matrix (.csv) as input. Meta data (.csv) is needed for the training process. The column cell_type in the meta data is used as the label for training. You can specify the model save path by using the argument <code class="docutils literal notranslate"><span class="pre">--model</span></code>. The model’s training epochs can be defined using the argument <code class="docutils literal notranslate"><span class="pre">--epochs</span></code>, and the model states will be saved at the end of each epoch. The training process duration can vary, depending on the size of your dataset and the computational capacity, and may range from minutes to several hours. Here is an example of the training process using the demo dataset.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">scGO</span><span class="o">.</span><span class="n">py</span> <span class="n">train</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="o">--</span><span class="n">gene_expression_matrix</span> <span class="n">GENE_EXPRESSION_MATRIX</span> <span class="p">[</span><span class="o">--</span><span class="n">meta_data</span> <span class="n">META_DATA</span><span class="p">]</span> <span class="o">--</span><span class="n">model</span> <span class="n">MODEL</span> <span class="p">[</span><span class="o">--</span><span class="n">task</span> <span class="n">TASK</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">epoch</span> <span class="n">EPOCH</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">batch_size</span> <span class="n">BATCH_SIZE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">learning_rate</span> <span class="n">LEARNING_RATE</span><span class="p">]</span>  <span class="p">[</span><span class="o">--</span><span class="n">label</span> <span class="n">LABEL</span><span class="p">]</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
    <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
    <span class="o">--</span><span class="n">gene_expression_matrix</span> <span class="n">GENE_EXPRESSION_MATRIX</span>      <span class="n">Train</span> <span class="n">data</span> <span class="n">file</span><span class="o">.</span>
    <span class="o">--</span><span class="n">meta_data</span> <span class="n">META_DATA</span>        <span class="n">Meta</span> <span class="n">data</span><span class="p">,</span> <span class="n">including</span> <span class="n">cell</span> <span class="nb">type</span> <span class="n">labels</span><span class="o">.</span>
    <span class="o">--</span><span class="n">model</span> <span class="n">MODEL</span>         <span class="n">Output</span> <span class="n">model</span> <span class="n">file</span><span class="o">.</span>
    <span class="o">--</span><span class="n">task</span> <span class="n">TASK</span>           <span class="n">Task</span> <span class="nb">type</span><span class="p">,</span> <span class="n">classification</span> <span class="ow">or</span> <span class="n">regression</span><span class="o">.</span>
    <span class="o">--</span><span class="n">epoch</span> <span class="n">EPOCH</span>         <span class="n">Number</span> <span class="n">of</span> <span class="n">training</span> <span class="n">epochs</span><span class="o">.</span>
    <span class="o">--</span><span class="n">batch_size</span> <span class="n">BATCH_SIZE</span>      <span class="n">Batch</span> <span class="n">size</span><span class="o">.</span>
    <span class="o">--</span><span class="n">learning_rate</span> <span class="n">LEARNING_RATE</span>       <span class="n">Learning</span> <span class="n">rate</span><span class="o">.</span>
    <span class="o">--</span><span class="n">label</span> <span class="n">LABEL</span>         <span class="n">Column</span> <span class="ow">in</span> <span class="n">meta_data</span> <span class="n">to</span> <span class="n">be</span> <span class="n">predicted</span><span class="o">.</span>

<span class="n">expample</span><span class="p">:</span>
<span class="n">python</span> <span class="n">scripts</span><span class="o">/</span><span class="n">scGO</span><span class="o">.</span><span class="n">py</span> <span class="n">train</span> <span class="o">--</span><span class="n">gene_expression_matrix</span> <span class="n">demo</span><span class="o">/</span><span class="n">baron_data_filtered</span><span class="o">.</span><span class="n">csv</span> <span class="o">--</span><span class="n">meta_data</span> <span class="n">demo</span><span class="o">/</span><span class="n">baron_meta_data</span><span class="o">.</span><span class="n">csv</span> <span class="o">--</span><span class="n">model</span> <span class="n">models</span><span class="o">/</span><span class="n">scGO</span><span class="o">.</span><span class="n">demo</span><span class="o">.</span><span class="n">pkl</span>
</pre></div>
</div>
</section>
<section id="training-regression-model-to-predict-a-continuous-cell-status">
<h2>Training regression model to predict a continuous cell status<a class="headerlink" href="#training-regression-model-to-predict-a-continuous-cell-status" title="Link to this heading"></a></h2>
<p>Set the <code class="docutils literal notranslate"><span class="pre">task</span></code> argument to <code class="docutils literal notranslate"><span class="pre">regression</span></code> and specify the <code class="docutils literal notranslate"><span class="pre">label</span></code> argument to correspond to a column in the metadata that you aim to predict.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">scripts</span><span class="o">/</span><span class="n">scGO</span><span class="o">.</span><span class="n">py</span> <span class="n">train</span> <span class="o">--</span><span class="n">gene_expression_matrix</span> <span class="n">demo</span><span class="o">/</span><span class="n">baron_data_filtered</span><span class="o">.</span><span class="n">csv</span> <span class="o">--</span><span class="n">task</span> <span class="n">regression</span> <span class="o">--</span><span class="n">epoch</span> <span class="mi">100</span> <span class="o">--</span><span class="n">batch_size</span> <span class="mi">8</span> <span class="o">--</span><span class="n">meta_data</span> <span class="n">demo</span><span class="o">/</span><span class="n">baron_meta_data_senescence_score</span><span class="o">.</span><span class="n">csv</span> <span class="o">--</span><span class="n">label</span> <span class="n">senescence_score</span> <span class="o">--</span><span class="n">model</span> <span class="n">models</span><span class="o">/</span><span class="n">scGO</span><span class="o">.</span><span class="n">senescence_score</span><span class="o">.</span><span class="n">demo</span><span class="o">.</span><span class="n">pkl</span>
</pre></div>
</div>
<p>To predict new data, load the pre-trained model, and generate predictions for the new data. The results should include both the predicted cell type label and the associated predicted value.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">scripts</span><span class="o">/</span><span class="n">scGO</span><span class="o">.</span><span class="n">py</span> <span class="n">predict</span> <span class="o">--</span><span class="n">gene_expression_matrix</span> <span class="n">demo</span><span class="o">/</span><span class="n">baron_data</span><span class="o">.</span><span class="n">csv</span> <span class="o">--</span><span class="n">task</span> <span class="n">regression</span> <span class="o">--</span><span class="n">model</span> <span class="n">models</span><span class="o">/</span><span class="n">scGO</span><span class="o">.</span><span class="n">senescence_score</span><span class="o">.</span><span class="n">demo</span><span class="o">.</span><span class="n">pkl</span> <span class="o">--</span><span class="n">output</span> <span class="n">demo</span><span class="o">/</span><span class="n">baron_meta_data_senescence_score</span><span class="o">.</span><span class="n">predicted</span><span class="o">.</span><span class="n">csv</span>
</pre></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="data_preprocessing.html" class="btn btn-neutral float-left" title="Data preprocessing" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="run_examples.html" class="btn btn-neutral float-right" title="Run examples" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Yulab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>